A51-custom-backend-metrics.md

```diff
@@ -0,0 +1,466 @@
+  // report interval separately. If the default report interval is not configured or badly configured 
+  // non-positive, the default is 1 min. 
+  // public BindableService createService(long minInterval, TimeUnit minUnit,
+  // long defaultInterval, TimeUnit defaultUnit)
+
+  // Creates an OOB bindable service using default minimum report interval = 30s.
+  public BindableService createService(ScheduledExecutorService timeService);
+
+  // Update the metrics value corresponding to the specified key.
+  public void setUtilizationMetric(String key, double value);
+
+  // Replace the whole metrics data using the specified map.
+  public void setAllUtilizationMetrics(Map<String, Double> metrics);
+
+  // Remove the metrics data entry corresponding to the specified key.
+  public void deleteUtilizationMetric(String key);
+
+  // Update the CPU utilization metrics data.
+  public void setCPUUtilizationMetric(double value);
+
+  // Clear the CPU utilization metrics data.
+  public void deleteCPUUtilizationMetric();
+
+  // Update the memory utilization metrics data.
+  public void setMemoryUtilizationMetric(double value);
+
+  // Clear the memory utilization metrics data.
+  public void deleteMemoryUtilizationMetric();
+}
+```
+
+At the client side, Java will provide a utility function to wrap the `LoadBalancer.Helper`,
+usable by both WRR policy and xDS LB policies. The `LoadBalancer.Helper` manages the OOB stream
+lifecycle and accepts subscriptions from any routing hierarchy.
+The OOB stream will also gracefully disable itself if the server integration is not complete, 
+so users do not have to worry about client and server deployment coordination.
+
+The following pseudocode shows how a custom load balancing policy instantiates the helper wrapper, 
+installs the out-of-band metrics reporting mechanism, and then consumes the metrics reports, 
+which only requires users to register an `OrcaOobReportListener` corresponding to the subchannel.
+
+```java
+class CustomLoadBalancer extends LoadBalancer { 
+  private final Helper orcaHelper;
+  
+  public CustomLoadBalancer(Helper helper) {
+    orcaHelper = OrcaUtil.newOrcaHelper(helper);
+  } 
+  
+  private Subchannel handleNewServer(ResolvedAddresses address) {
+    Subchannel subchannel = orcaHelper.createSubchannel(
+        CreateSubchannelArgs.newBuilder().setAddresses(address).build());
+    // Create a separate listener for each Subchannel.
+    OrcaOobReportListener listener = new OrcaOobReportListenerImpl(subchannel);
+    OrcaUtil.setListener(subchannel, listener, 
+            OrcaReportingConfig.newBuilder().setInterval(40, SECONDS).build());
+    return subchannel;
+  }
+}
+
+// Users implement a listener and pass it to the client library to receive
+// callbacks upon metric report updates from the backend.
+public interface OrcaOobReportListener {
+  //receive load report in the format of ORCA protocol.
+  void onLoadReport(OrcaLoadReport report);
+}
+
+```
+
+### Go
```

Comment:
I think we can just delete the Go section and have it be created when there is content to include.

Reply:
SGTM.

Generally, should we replace all these sections with pointers to the implementation / API documentation once done?

Reply:
It looks either way exists, with pseudo-code + description pattern more popular among proposals.

Reply:
Generally that sort of information I add to the implementation section; I would have put this Java/C stuff in the Proposal section. But part of this strangeness is because our gRFCs have started including a lot more per-language information instead of saying, "these are the requirements; implement it how you desire" which applies equally to all languages.

https://github.com/grpc/proposal/blob/master/L74-java-channel-creds.md#implementation
https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md#implementation (note: I added the C mapping because I had repeated difficulty figuring it out)
https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md#implementation

Reply:
IMO we should scrub any language implementation stuff (except maybe high level design concepts) from gRFCs after the implementation is complete.  It's more likely to be inaccurate than the javadoc/godoc/etc, since: 1. the actual implementation (v1) may have needed to diverge from the initial gRFC, and 2. the implementation may change over time (new methods or fields are added, etc), but it is rare anyone will think to revisit the gRFC to update the documentation.

Reply:
I think it's fine to keep this here.  The APIs are part of the design here.  They will evolve over time, but so will the design itself.  This doc just shows the design at the time that we finish working on it; it's not meant to be a living doc that we always keep up to date.